src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -1432,12 +1399,11 @@ private async ValueTask<(Stream?, HttpResponseMessage?)> EstablishProxyTunnelAsy
 
             if (tunnelResponse.StatusCode != HttpStatusCode.OK)
             {
-                return (null, tunnelResponse);
+                tunnelResponse.Dispose();
+                throw new HttpRequestException(SR.Format(SR.net_http_proxy_tunnel_returned_failure_status_code, _proxyUri, (int)tunnelResponse.StatusCode));
             }
 
-            Stream stream = tunnelResponse.Content.ReadAsStream(cancellationToken);
-
-            return (stream, null);
+            return tunnelResponse.Content.ReadAsStream(cancellationToken);
```

Comment:
If this ReadAsStream gets canceled (which I think would only happen if cancellation were already requested when it's called, since it's otherwise synchronous and immediate), do we need to Dispose of tunnelResponse?

Reply:
We probably should, yeah. I'll add some logic for this.

Reply:
I suppose technically we should be using the async overload here as well, at least when it is async...

Reply:
We could, but we'd changed these to use the sync one as a) we own the response content and know its behavior and b) the async one results in an unnecessary allocation.  From my perspective, there should never have been a ReadAsStreamAsync... it's never async :)

Reply:
Should we obsolete it, then?

(And why does the sync version take a CancellationToken?)

Reply:
> Should we obsolete it, then?

Everyone uses it.  I think it'd be too painful, and in reality it's only negatively impactful if you're hyperoptimizing.  I think we have bigger fish to fry... cough...headers...cough.

> And why does the sync version take a CancellationToken?

Good question.  I'm 99% sure I commented on it being unnecessary.  I'd need to look at the code, but if memory serves technically the base implementation could actually do I/O (making the XxAsync version still relevant) if it's not overridden and delegates to the CreateContentReadStream{Async} implementation.